aea/test_tools/test_skill.py

```diff
@@ -0,0 +1,345 @@
+            for any other message, it is the opposite of the one preceding it.
+         - 'target' is integer, representing the target of the message.
+            - 'target' is optional, if not provided: for the first message is assumed 0,
+            for any other message, it is the index of the message before it in the tuple of messages + 1.
+         - 'performative' represents the performative of the message.
+         - contents is dictionary of the message's contents.
+
+        :param dialogues: a dialogues class
+        :param counterparty: the message_id
+        :param messages: the dialogue_reference
+
+        :return: the created incoming message
+        """
+        incomplete_message_1 = messages[0]
+        complete_compact_message_1 = self._complete_compact_message(
+            incomplete_message_1, last_incoming=True, message_id=1,
+        )
+        last_incoming = complete_compact_message_1[0]
+        is_incoming = complete_compact_message_1[0]
+        message_id = messages.index(incomplete_message_1) + 1
+        target = complete_compact_message_1[1]
+        performative = complete_compact_message_1[2]
+        contents = complete_compact_message_1[3]
+
+        if is_incoming:  # messages from the opponent
+            dialogue_reference = dialogues.new_self_initiated_dialogue_reference()
+            message = self.build_incoming_message(
+                message_type=dialogues._message_class,  # pylint: disable=protected-access
+                dialogue_reference=dialogue_reference,
+                message_id=message_id,
+                target=target,
+                performative=performative,
+                to=self.skill.skill_context.agent_address,
+                sender=counterparty,
+                **contents,
+            )
+            dialogue = cast(Dialogue, dialogues.update(message))
+            if dialogue is None:
+                raise AEAEnforceError(
+                    "Cannot update the dialogue with message number {}".format(
+                        message_id
+                    )
+                )
+        else:  # messages from self
+            _, dialogue = dialogues.create(
+                counterparty=counterparty, performative=performative, **contents
+            )
+
+        for incomplete_message in messages[1:]:  # type: ignore
+            complete_compact_message = self._complete_compact_message(
+                incomplete_message,
+                last_incoming=last_incoming,
+                message_id=messages.index(incomplete_message) + 1,
+            )
+            last_incoming = complete_compact_message[0]
+            is_incoming = complete_compact_message[0]
+            message_id = messages.index(incomplete_message) + 1
+            target = complete_compact_message[1]
+            performative = complete_compact_message[2]
+            contents = complete_compact_message[3]
```

Comment:
what's the point of `_complete_compact_message` if you then need another 6 lines to unpack it?

Reply:
This method gives a semantics to the unspecified `target` and `is_incoming`. 
The former becomes `last_message_id +1` and the latter becomes the opposite of the preceding message.

Reply:
Maybe a bad name. I can also make it so unpacking is easier

Reply:
I refactored them, so hopefully it's more clear.